File: tools/java/OperationTrace.java
package blockforge.tools;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.EnumMap;
import java.util.List;
import java.util.Locale;

/** Decoder and validator for filesystem_operation_fuzzer byte sequences. */
public final class OperationTrace {
    private static final int MAX_OPERATIONS = 4096;

    private OperationTrace() {
    }

    public enum Kind {
        MKDIR,
        CREATE,
        WRITE,
        APPEND,
        READ,
        REMOVE,
        RENAME,
        HARD_LINK,
        SYMLINK,
        LIST,
        WALK,
        BEGIN,
        ROLLBACK,
        COMMIT,
        CHECKPOINT,
        SET_XATTR,
        GET_XATTR,
        REMOVE_XATTR,
        SERIALIZE_REOPEN,
        STAT,
        RECURSIVE_REMOVE;

        static Kind fromByte(int value) {
            return values()[value % values().length];
        }

        boolean mutates() {
            return switch (this) {
                case MKDIR, CREATE, WRITE, APPEND, REMOVE, RENAME,
                     HARD_LINK, SYMLINK, SET_XATTR, REMOVE_XATTR,
                     RECURSIVE_REMOVE -> true;
                default -> false;
            };
        }

        boolean reads() {
            return switch (this) {
                case READ, LIST, WALK, GET_XATTR, STAT -> true;
                default -> false;
            };
        }
    }

    public record Operation(
        int ordinal,
        int raw,
        Kind kind,
        int slot,
        int value,
        int transactionDepth
    ) {
    }

    public record Report(
        Path path,
        int bytes,
        List<Operation> operations,
        List<String> warnings,
        int maximumTransactionDepth
    ) {
        Report {
            operations = List.copyOf(operations);
            warnings = List.copyOf(warnings);
        }

        long mutationCount() {
            return operations.stream()
                .filter(operation -> operation.kind().mutates())
                .count();
        }

        long readCount() {
            return operations.stream()
                .filter(operation -> operation.kind().reads())
                .count();
        }

        EnumMap<Kind, Integer> counts() {
            EnumMap<Kind, Integer> result = new EnumMap<>(Kind.class);
            for (Operation operation : operations) {
                result.merge(operation.kind(), 1, Integer::sum);
            }
            return result;
        }
    }

    public static Report decode(Path path) throws IOException {
        byte[] data = Files.readAllBytes(path);
        int count = Math.min(data.length, MAX_OPERATIONS);
        List<Operation> operations = new ArrayList<>(count);
        List<String> warnings = new ArrayList<>();
        int transactionDepth = 0;
        int maximumDepth = 0;
        for (int index = 0; index < count; index++) {
            int raw = data[index] & 0xff;
            Kind kind = Kind.fromByte(raw);
            if (kind == Kind.BEGIN) {
                if (transactionDepth != 0) {
                    warnings.add(
                        "operation " + index + ": nested BEGIN"
                    );
                }
                transactionDepth++;
                maximumDepth = Math.max(maximumDepth, transactionDepth);
            } else if (kind == Kind.COMMIT || kind == Kind.ROLLBACK) {
                if (transactionDepth == 0) {
                    warnings.add(
                        "operation " + index + ": " + kind
                            + " without BEGIN"
                    );
                } else {
                    transactionDepth--;
                }
            } else if (kind == Kind.CHECKPOINT && transactionDepth != 0) {
                warnings.add(
                    "operation " + index
                        + ": CHECKPOINT inside transaction"
                );
            }
            operations.add(
                new Operation(
                    index,
                    raw,
                    kind,
                    (raw >>> 3) & 31,
                    (raw * 257 + index * 17) & 0xffff,
                    transactionDepth
                )
            );
        }
        if (data.length > MAX_OPERATIONS) {
            warnings.add(
                (data.length - MAX_OPERATIONS)
                    + " operations omitted by resource limit"
            );
        }
        if (transactionDepth != 0) {
            warnings.add(
                "sequence ends with " + transactionDepth
                    + " active transaction(s)"
            );
        }
        return new Report(
            path,
            data.length,
            operations,
            warnings,
            maximumDepth
        );
    }

    private static void print(Report report, boolean verbose) {
        System.out.printf(
            Locale.ROOT,
            "%s: bytes=%d operations=%d mutations=%d reads=%d "
                + "maximum-transaction-depth=%d%n",
            report.path(),
            report.bytes(),
            report.operations().size(),
            report.mutationCount(),
            report.readCount(),
            report.maximumTransactionDepth()
        );
        for (Kind kind : Kind.values()) {
            int count = report.counts().getOrDefault(kind, 0);
            if (count != 0) {
                System.out.printf(
                    Locale.ROOT,
                    "  %-20s %d%n",
                    kind,
                    count
                );
            }
        }
        for (String warning : report.warnings()) {
            System.out.println("  warning: " + warning);
        }
        if (verbose) {
            for (Operation operation : report.operations()) {
                System.out.printf(
                    Locale.ROOT,
                    "  %4d raw=%3d %-20s slot=%2d value=%5d tx=%d%n",
                    operation.ordinal(),
                    operation.raw(),
                    operation.kind(),
                    operation.slot(),
                    operation.value(),
                    operation.transactionDepth()
                );
            }
        }
    }

    private static void usage() {
        System.err.println("Usage: OperationTrace [--verbose] SEED...");
    }

    public static void main(String[] arguments) {
        boolean verbose = false;
        List<Path> paths = new ArrayList<>();
        for (String argument : arguments) {
            if (argument.equals("--verbose")) {
                verbose = true;
            } else if (argument.startsWith("-")) {
                usage();
                System.exit(2);
            } else {
                paths.add(Path.of(argument));
            }
        }
        if (paths.isEmpty()) {
            usage();
            System.exit(2);
        }
        try {
            for (Path path : paths) {
                print(decode(path), verbose);
            }
        } catch (IOException error) {
            System.err.println("OperationTrace: " + error.getMessage());
            System.exit(1);
        }
    }
}

File: tools/python/operation_trace.py
#!/usr/bin/env python3
"""Decode, generate, and compare BlockForge operation-fuzzer sequences."""

from __future__ import annotations

import argparse
import dataclasses
import json
import random
import sys
from collections import Counter
from pathlib import Path
from typing import Iterable

MAX_OPERATIONS = 4096

OPERATIONS = (
    "mkdir",
    "create",
    "write",
    "append",
    "read",
    "remove",
    "rename",
    "hard-link",
    "symlink",
    "list",
    "walk",
    "begin",
    "rollback",
    "commit",
    "checkpoint",
    "set-xattr",
    "get-xattr",
    "remove-xattr",
    "serialize-reopen",
    "stat",
    "recursive-remove",
)


@dataclasses.dataclass(frozen=True)
class Operation:
    ordinal: int
    opcode: int
    name: str
    slot: int
    value: int
    transaction_depth: int


def decode(data: bytes) -> tuple[list[Operation], list[str]]:
    operations: list[Operation] = []
    warnings: list[str] = []
    depth = 0
    count = min(len(data), MAX_OPERATIONS)
    for index, byte in enumerate(data[:count]):
        opcode = byte % len(OPERATIONS)
        name = OPERATIONS[opcode]
        if name == "begin":
            if depth:
                warnings.append(f"operation {index}: nested transaction")
            depth += 1
        elif name in {"rollback", "commit"}:
            if not depth:
                warnings.append(
                    f"operation {index}: {name} without transaction"
                )
            else:
                depth -= 1
        operations.append(
            Operation(
                ordinal=index,
                opcode=opcode,
                name=name,
                slot=(byte >> 3) & 31,
                value=(byte * 257 + index * 17) & 0xFFFF,
                transaction_depth=depth,
            )
        )
    if len(data) > MAX_OPERATIONS:
        warnings.append(
            f"{len(data) - MAX_OPERATIONS} operations omitted by limit"
        )
    if depth:
        warnings.append(f"sequence ends with {depth} active transaction(s)")
    return operations, warnings


def report(path: Path) -> dict[str, object]:
    data = path.read_bytes()
    operations, warnings = decode(data)
    counts = Counter(operation.name for operation in operations)
    maximum_depth = max(
        (operation.transaction_depth for operation in operations), default=0
    )
    mutating = {
        "mkdir",
        "create",
        "write",
        "append",
        "remove",
        "rename",
        "hard-link",
        "symlink",
        "recursive-remove",
        "set-xattr",
        "remove-xattr",
    }
    return {
        "path": str(path),
        "bytes": len(data),
        "operations": len(operations),
        "mutations": sum(operation.name in mutating for operation in operations),
        "reads": sum(
            operation.name
            in {"read", "list", "walk", "get-xattr", "stat"}
            for operation in operations
        ),
        "maximum_transaction_depth": maximum_depth,
        "counts": dict(sorted(counts.items())),
        "warnings": warnings,
        "trace": [dataclasses.asdict(operation) for operation in operations],
    }


def render_text(value: dict[str, object], verbose: bool) -> str:
    lines = [
        f"{value['path']}:",
        f"  bytes: {value['bytes']}",
        f"  operations: {value['operations']}",
        f"  mutations: {value['mutations']}",
        f"  reads: {value['reads']}",
        f"  maximum transaction depth: {value['maximum_transaction_depth']}",
        "  operation counts:",
    ]
    for name, count in value["counts"].items():  # type: ignore[union-attr]
        lines.append(f"    {name}: {count}")
    warnings = value["warnings"]
    if warnings:
        lines.append("  warnings:")
        lines.extend(f"    {warning}" for warning in warnings)  # type: ignore[arg-type]
    if verbose:
        lines.append("  trace:")
        for operation in value["trace"]:  # type: ignore[assignment]
            lines.append(
                f"    {operation['ordinal']:4} {operation['name']:<18} "
                f"slot={operation['slot']:<2} value={operation['value']:<5} "
                f"tx-depth={operation['transaction_depth']}"
            )
    return "\n".join(lines)


def generate(length: int, seed: int) -> bytes:
    if length < 0 or length > MAX_OPERATIONS:
        raise ValueError(f"length must be between 0 and {MAX_OPERATIONS}")
    generator = random.Random(seed)
    output = bytearray()
    transaction = False
    for index in range(length):
        if index % 31 == 0 and not transaction:
            opcode = OPERATIONS.index("begin")
            transaction = True
        elif index % 31 == 30 and transaction:
            opcode = generator.choice(
                (OPERATIONS.index("commit"), OPERATIONS.index("rollback"))
            )
            transaction = False
        else:
            opcode = generator.randrange(len(OPERATIONS))
            if transaction and OPERATIONS[opcode] == "begin":
                opcode = OPERATIONS.index("write")
            if not transaction and OPERATIONS[opcode] in {"commit", "rollback"}:
                opcode = OPERATIONS.index("read")
        upper = generator.randrange(0, 256 // len(OPERATIONS) + 1)
        output.append((opcode + upper * len(OPERATIONS)) & 0xFF)
    if transaction and output:
        output[-1] = OPERATIONS.index("rollback")
    return bytes(output)


def _parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    subparsers = parser.add_subparsers(dest="command", required=True)
    inspect = subparsers.add_parser("inspect")
    inspect.add_argument("seeds", type=Path, nargs="+")
    inspect.add_argument("--json", action="store_true")
    inspect.add_argument("--verbose", action="store_true")
    create = subparsers.add_parser("generate")
    create.add_argument("output", type=Path)
    create.add_argument("--length", type=int, default=256)
    create.add_argument("--seed", type=int, default=1)
    return parser


def main(argv: Iterable[str] | None = None) -> int:
    arguments = _parser().parse_args(argv)
    try:
        if arguments.command == "generate":
            arguments.output.parent.mkdir(parents=True, exist_ok=True)
            arguments.output.write_bytes(
                generate(arguments.length, arguments.seed)
            )
            print(f"wrote {arguments.output}")
            return 0
        reports = [report(path) for path in arguments.seeds]
        if arguments.json:
            print(json.dumps(reports, indent=2))
        else:
            print(
                "\n\n".join(
                    render_text(value, arguments.verbose)
                    for value in reports
                )
            )
        return 0
    except (OSError, ValueError) as error:
        print(f"operation_trace: {error}", file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())
